src/HttpReply.cc

```diff
@@ -82,18 +82,27 @@ HttpReply::clean()
 }
 
 void
-HttpReply::packHeadersInto(Packable * p) const
+HttpReply::packHeadersIntoFast(Packable &p) const
```

Comment:
This naming does not look right. The "fast" variant is the existing behaviour and the 'optimized' behaviour is "slow".

The difference in these methods is actually buffering (or lack thereof) in the HttpReply object. So please s/Fast/Unbuffered/  s/Slow/Buffered/

Reply:
>  The "fast" variant is the existing behaviour and the 'optimized' behaviour is "slow".

To be more precise, I assume you suggest s/Fast/Buffered/ s/Slow/Unbuffered/ (because 'slow' version does extra buffering step itself).
The fact that the existing packHeadersInto() became "fast" without changes, as you noted, has another sense: each of callers now is responsible for the decision of selection between 'fast' and 'slow'. It should use 'fast' version only if it "can quickly process small additions" (as method description says), and 'slow' otherwise. I don't see a big difference between fast/slow and buffered/unbuffered terminology, because probably they are equivalent. However fast/slow wording is slightly better because the code explicitly marks out cases need future optimization.

Reply:
I mean "existing behaviour" in sense that the packHeadersIntoFast() contains *exactly the same code* which Squid is doing right now. The diff hides it, but there is only a renaming of that method to say "Fast" and addition of one named as "Slow" optimized to avoid the problems in certain Packable's implementation.

Since the speed is due to the type of Packable child it is easy for callers to pass anything like StoreEntry to the "Fast" method and its suddenly worse than using "Slow". Despite "Fast" being super attractive when speed of producing a reply (eg StoreEntry) is most important.

Those Using/Assuming-FastPacker suggestions have the same problem. It's the word "fast" and "slow" themselves being deceptive/wrong here because HttpReply has zero control over the speed aspect (and that is fine). A Packable is a Packable and the callers (or rather dev working with callers) of HttpReply should not even have to be aware of their internal design.


> P.S. Changing the packer API to distinguish fast and slow packers at runtime is not a good idea
> because whether the packer is "fast enough" or "slow enough" depends (in part) on what is being
> packed. The caller should make that call when it can. All of the current callers can do that
> according to this PR. Besides, calling another virtual method is an extra expense that none of the
> current callers need to pay.

That is why I am *not* asking for a change to the API design. Just a rename of the methods to use less confusing words. I'm not completely set on "buffered", just avoiding the fast vs slow confusions.


PPS. the real problem for StoreEntry is bug 4477 and/or that its Packable::buffer()/flush() implementation is insufficient to protect it from small appends(). For ShmWriter its the Packable::buffer()/flush() methods being inherited is essentially a promise that ShmWriter Packable's are not affected by small writes. Since ShmWriter apparently _are_ affected (according to the PR description) the missing override methods should be added.

Perhapse fixing those problems should happen instead of this? that would certainly improve baseline performance. Even if we had to come back to this on top of the fixes, we would then know that this is truly useful change instead of having it silently become a pointless bit of complexity when those bugs are fixed.

Reply:
> HttpReply has zero control over the speed aspect

That is only correct in a myopic view that sees nothing but a slowPacker->pack() call inside the packing loop. When one looks at the IntoSlowPacker() method as a whole, it is clear that the method can speed up packing into a slow packer considerably.

> it is easy for callers to pass anything like StoreEntry to the "Fast" method and its suddenly worse than using "Slow".

Agreed. The MemBuf/Store-based names proposed below partially address this problem (at the expense of making the methods less general).

Certainly, using "Buffered" is even worse (i.e., more confusing) in that respect as have been already illustrated at the [beginning](https://github.com/squid-cache/squid/pull/230#discussion_r199316882) of this comment thread -- it is a lot more likely that the developers and reviewers will not know whether a "buffered" method should be used than they will misidentify the "fast" packer.

> the real problem for StoreEntry is bug 4477

The proposed optimization would remain useful even after bug 4477 is fixed -- even a very fast search through the memory nodes is still fairly expensive for adding ~200 items per message. This is a "optimization level" problem: Those 200 Store calls should be handled at higher level because any low-level Store optimization would not be able to achieve a similar performance boost.

> ... and/or that its Packable::buffer()/flush() implementation is insufficient to protect it from small appends().

That may be true. StoreEntry::buffer() currently protects only a portion of Store operations.  Investigating whether that protection can and should be extended to memory appends (by adding a temporary buffer to MemObject?) is a good idea. Another option is to implement this buffering inside (slow) Packers.

> Perhapse fixing those problems should happen instead of this?

Yes, it is possible that improving Packable::buffer() protection for StoreEntry would be overall better. Unfortunately, I cannot offer those improvements, so, assuming you are not going to fix those problems soon either, we have to pick among these two options:

1. Veto this PR (for some strange reason that I personally cannot think of).
2. Accept this PR (with some method names that we all can live with).

@yadij, your call. If you decide to move towards acceptance, I can offer these naming options:

For fast packers:

1. packHeadersIntoFastPacker()
1. packHeadersIntoMemBuf()
1. packHeadersIntoPackerWithFastBufferedWrites()
1. packHeadersIntoPackerThatBuffersWell()

For slow packers:

1. packHeadersIntoSlowPacker()
1. packHeadersIntoStore()
1. packHeadersIntoPackerWithSlowBufferedWrites()
1. packHeadersIntoPackerThatBuffersPoorly()

where `Into` can also be `Using` or `Assuming` when appropriate.

Personally, I can live with any of the above. FWIW, they are listed in a rough order of my preference.

Reply:
@yadij, should I leave methods naming as is or rename as suggested?

Reply:
Since this is a style issue and Alex and I disagree its up to you s implementer to make the call. If you wish to leave as-is despite the discussion above I will remove my request, or approve whatever naming patch you apply.

Reply:
@yadij, I've corrected method names at fa34c9b.